Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,314 @@
+---
+title: Implicitly-scoped qubit allocation
+description: Provides more convenient syntax for allocating qubits.
+author: Sarah Marshall
+date: TBD
+---
+
+# Proposal
+
+1. Rename `using` and `borrowing` keywords to `use` and `borrow`.
+2. Allow `use` and `borrow` statements to be followed by either a block `{ ... }` or a semicolon `;`.
+3. Remove the requirement for parentheses around headers for all block statements.
```

Comment:
Is the intent to make the parenthesis permanently optional, or to ultimately deprecate their use (i.e. error on use)?

Reply:
The intent is to make parentheses a syntax error, except in cases where the parentheses contain a single expression, in which case it will continue to be supported since any expression can be wrapped in parentheses.

Reply:
I would prefer to have the parentheses optional at least as it makes the code easier to read for people like me coming from a C++/C# background. Was this decided based on user data?

Reply:
Could you clarify how

```qsharp
for (x in xs) {
   // ...
}
```

is more readable than

```qsharp
for x in xs {
    // ...
}
```

?

Reply:
About making them optional, I would prefer not to have too many ways to express the same construct in Q#, so I think we should pick one or the other rather than supporting both. If the non-parenthesized form has an issue with readability, then I don't think making it optional will fully solve that problem because code written by other people could still choose not to use parentheses.

Reply:
If we want to go one way versus the other permanently on account of readability, then we should make a decision on what is more readable for our users based on user data, rather than just our perspectives. I'm tempted to defend the readability of parentheses here in an answer to your question above, but I think that would undermine my point of using user data 😊.

Reply:
The parentheses were removed for consistency with the newly added non-block `use` and `borrow` statements. Since we now have

```qsharp
use q = Qubit();
```

It's consistent for the block form to be

```qsharp
use q = Qubit() {
    // ...
}
```

instead of

```qsharp
use (q = Qubit()) {
    // ...
}
```

We would need an argument that removing the parentheses has some cost that is greater than the benefit gained from consistent syntax. I don't think user data by itself is able to provide that argument.

Reply:
Since I don't see any significant difference in readability, I am fine with making this call here. Allowing the parenthesis to me doesn't add enough value to justify the existence of a redundant way of expressing the same concept (see also our [guidelines](https://github.com/microsoft/qsharp-language/blob/main/Guidelines.md)). For consistency reasons I agree with dropping the parentheses.